SQA_scraper (issue #136)

nutcracker22 opened the pull request:
@ormiret, @JackGilmore, @KarenJewell
This is the first draft of the scraper for the Scottish Qualifications Authority. It currently lacks information about the licences, since I did not find anything about it on their webpage. I'll write them an email and ask about the licences.

Are there any changes you would like me to apply?

Btw, the scraper is not yet called in Github actions, so it needs to be added there or in main.py (depending on what is added first to the main branch).

KarenJewell commented:
looks great!

3 things from me:
1. `Description` is empty although, I'm not really sure what we can put in it either!
2. `DateCreated` is rendering weird for me when viewing the output csv in excel. Might need treatment here.
3. I'm not sure of the dataset titles 🙃 the way you've got it will generate 639 distinct datasets (i.e. one ODS listing per file) and it is technically accurate. But now I'm not sure what a grouping should be. SQA have it as multiple qualifications per year (i.e. 1 year = 1 dataset, each qual is an asset), but I wonder if the more intuitive way of grouping it should be by qualification (i.e. 1 qualification = 1 dataset, each year is an asset) this leans a bit into a definition of what a "dataset" is. Hypothetically, if a publisher adds a new year of data - is that a new dataset or an extension to an existing dataset?

nutcracker22 commented:
So far I've been mainly thinking about 3). The grouping according to qualification sounds to me definitely like the most viable option. However, am I getting it right, that currently [export2jkan.py, line 86](https://github.com/OpenDataScotland/the_od_bods/blob/main/export2jkan.py#L86) defines a dataset, when both the page_url and the title match? Thus, for this source this approach would not work, since page_url for each year differs from the other years for the same title. Means, this would require is workaround depending on the owner.

If that is fine with you, I can try to code it. Otherwise, grouping datasets according to year, as done on the SQA webpage, is the alternative.

I will also look into your first and second point.

KarenJewell commented:
Catching up on outstanding items - you're right, JKAN groups datasets by pageURL + title combo. The JKAN base assumption being that a dataset is one set of values in multiple file formats. Whereas we're seeing different interpretations from publishers also, e.g. the choice to split a large dataset into multiple volumes is very valid.

I think for now, grouping by year as SQA have done makes best progress. The "right" approach takes us to a whole bigger discussion about what defines a dataset.

nutcracker22 commented:
@KarenJewell, I changed the title and added a (very general) description to fit the grouping by year. This will create 26 datasets (if I counted correctly), one for each year, and many download links for each year. This should match what we discussed above.

In case, that these are too many links for each dataset, I just got the idea, that we could use the `<h2>` headings (like 'Annual Statistical Report', 'Progression Tables 2021' and so on) to create more datasets for each year. But have a look at it, first.

Is the `DateCreated` still rendering weird?

KarenJewell commented on web-scrapers/qualifications.json:
```diff
@@ -0,0 +1,19 @@
+{"qualifications":[
```
qualifications.json doesn't seem to be used anywhere. Is this meant for something? If not it could be good to include in the dataset description if this is being extracted from somewhere?

nutcracker22 replied:
it was used in a previous commit, when we still had each file as a dataset, to create the descriptions for the datasets. The information in this file is not scraped, but only copied from the SQA website.

However, now the datasets are different, so I don't know how to use it anymore. I would delete the file from the PR, unless we find a good use for it.

KarenJewell replied:
agree with delete.

KarenJewell commented on web-scrapers/sqa_scraper.py:
```diff
@@ -0,0 +1,276 @@
+# Packages: beautifulsoup4, csv, requests, math
+import requests
+import csv
+import json
+import re
```
Doesn't seem these 2 packages are being used.

KarenJewell commented on web-scrapers/sqa_scraper.py:
```diff
@@ -0,0 +1,276 @@
+    print("Getting available years")
+    category_links = fetch_available_years()
+    for year_string in category_links.keys():
+        print("Getting", year_string)
+        years_page = fetch_year_page(category_links[year_string])
+        owner = "Scottish Qualifications Authority (SQA)"
+        title = create_title(year_string)
+        # print("title", title)
+        description = fetch_description(years_page, year_string)
+        print(description)
```
scraper is very verbose. I would recommend removing the print statement for `description` as it's not any more information than the print statement in line 225.

KarenJewell reviewed (changes requested):
Looking great! A few things left
- the pipeline is missing the piece which integrates it in `merge_data.py`
- merge_data.py uses a `filename` attribute which we can use an extract of `AssetURL` to do.
- would recommend going through and removing commented debug statements

It is very nearly there - most are minor styling comments but the one major thing needed before merge is the integration into `merge_data.py` piece. Will be worth seeing what the other scrapers have done in it.

nutcracker22 commented:
Hi @KarenJewell, I think I addressed all the issues you rightfully raised. For the pipeline I will create a separate PR to include the new SQA scraper :)

KarenJewell commented:
Looks fantastic. Thanks for this @nutcracker22!
